drivers/linstorvolumemanager.py

```diff
@@ -196,18 +200,65 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
-        uri = _get_controller_uri()
-        if uri:
-            return uri
 
-        retries += 1
-        if retries >= 10:
-            break
-        time.sleep(1)
+def get_cached_controller_uri():
+    try:
+        with open(CONTROLLER_CACHE_PATH, "r") as f:
+            return f.read().strip()
+    except (PermissionError, IsADirectoryError, FileNotFoundError) as e:
+        util.SMlog("Unable to read controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e),
+                   priority=util.LOG_DEBUG)
+    except OSError as e:
+        util.SMlog("OS error reading controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e))
+    except Exception as e:
+        util.SMlog('Unable to read controller uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH,e))
+    return None
 
+def delete_controller_uri_cache():
+    try:
+        os.remove(CONTROLLER_CACHE_PATH)
+    except (PermissionError, IsADirectoryError, FileNotFoundError) as e:
+        util.SMlog("Unable to delete controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e),
+                   priority=util.LOG_DEBUG)
+    except OSError as e:
+        util.SMlog("OS error deleting controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e))
+    except Exception as e:
+        util.SMlog('Unable to delete uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
+
+def write_controller_uri_cache(uri):
+    try:
+        if not os.path.exists(CONTROLLER_CACHE_DIRECTORY):
+            os.makedirs(CONTROLLER_CACHE_DIRECTORY)
+            os.chmod(CONTROLLER_CACHE_DIRECTORY, 0o700)
+        with open(CONTROLLER_CACHE_PATH, "w") as f:
+            f.write(uri)
+    except (PermissionError, IsADirectoryError, FileNotFoundError) as e:
+        util.SMlog("Unable to write controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e),
+                   priority=util.LOG_DEBUG)
+    except OSError as e:
+        util.SMlog("OS error writing controller uri cache file at `{}`: {}".format(CONTROLLER_CACHE_PATH, e))
+    except Exception as e:
+        util.SMlog('Unable to write uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
```

Comment:
```suggestion
def write_controller_uri_cache(uri):
    try:
        if not os.path.exists(CONTROLLER_CACHE_DIRECTORY):
            os.makedirs(CONTROLLER_CACHE_DIRECTORY)
            os.chmod(CONTROLLER_CACHE_DIRECTORY, 0o700)
        with open(CONTROLLER_CACHE_PATH, "w") as f:
            f.write(uri)
    except FileExistsError as e:
        pass
    except Exception as e:
        util.SMlog('Unable to write URI cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
```

Of course this suggestion is not totally robust in case of lock concurrency.

Reply:
Applied this in [0f8abe9](https://github.com/xcp-ng/sm/pull/83/commits/0f8abe9259520e340ad907b65a02f6bd7c7269d9)

Lock mechanism is in [6b816c3](https://github.com/xcp-ng/sm/pull/83/commits/6b816c3a2a3ac5e291c2e15ea6e62849d5935ce9)